Simple text example

koober opened the pull request:
added basic text example
added link to example in docs/api/extras/core/Font.html
added link to example in docs/api/loaders/FontLoader.html

koober committed:
Create webgl_geometry_text_shapes.html

koober committed:
Merge pull request #1 from koober/koober-text-example

Create webgl_geometry_text_shapes.html

koober committed:
Add files via upload

koober committed:
Add files via upload

koober commented:
Ref Issue #10725

WestLangley commented on examples/webgl_geometry_text_shapes.html:
```diff
@@ -0,0 +1,162 @@
+
+			function xMove( shape, shapeMid ) {
+
+				return shape.applyMatrix( new THREE.Matrix4().makeTranslation( shapeMid, 0, 0 ) );
```
```javascript
return shape.translate( shapeMid, 0, 0 );
```
You might as well inline that method.

koob3d replied:
Neat! Will do... thanks.

WestLangley commented:
https://rawgit.com/koober/three.js/dev/examples/webgl_geometry_text_shapes.html

Nice 2D text example. Any idea why the example has 27 geometries and 26 draw calls?

koob3d commented:
Confession. I only use  console.log(renderer.info.render.calls), which says 27.
Suggestions for a better stats tool welcome.

koober committed:
update xmove function

Moved and modded as westLangley proposed

WestLangley commented:
This issue is not with the stats tool, unless it has a bug.

Why so many draw calls, why so many geometries, and why fewer draw calls than geometries?

WestLangley commented on examples/webgl_geometry_text_shapes.html:
```diff
@@ -0,0 +1,162 @@
+							for ( var j = 0; j < shape.holes.length; j ++ ) {
+
+								var hole = shape.holes[ j ];
+								shapes.push( hole );
```
You are modifying the length of shapes as you iterate over it.

WestLangley commented on examples/webgl_geometry_text_shapes.html:
```diff
@@ -0,0 +1,162 @@
+						side: THREE.DoubleSide
+					} );
+
+					var xMove = function( shape, shapeMid ) {
```
Like I said, you don't use the return value and you don't need a separate function for this; just inline the code.

WestLangley commented on examples/webgl_geometry_text_shapes.html:
```diff
@@ -0,0 +1,162 @@
+
+						xMove( lineGeometry, xMid );
+
+						var lineMesh = new THREE.Line( lineGeometry, matDark );
```
`Line` uses `LineBasicMaterial`, not `MeshBasicMaterial`.

koober committed:
Remove xMove function, tidy shapes array size

koober committed:
Correct material for Line

koober commented:
Yes, I see what he meant by hacky now.
Sorry, I didn't wish to imply fault in any stats tool. 
For a single drawcall on the line geometries, am I right to think they need to be converted to LineSegments and be merged? I can play with that for a while but not too sure of a result. 

Or I can remove the line text? Then rather have a full example it should just go in the docs page for Font, which doesn't have an example attributed to it currently

WestLangley commented:
@koober I will leave it up to @mrdoob. Options are to (1) leave it as it is, (2) remove the line text, (3) do (2) and remove the example altogether, adding it to the docs page for Font, or (4) merge the line geometry. `BufferGeometry.merge()` works, but you have to pre-allocate the buffers.

`Geometry` merge is really inefficient for a large number of geometries. We could use this as an opportunity to show how to merge buffer geometries.

mrdoob commented:
Maybe we could even turn this into `THREE.OutlineShapeGeometry`...?

WestLangley commented:
>Maybe we could even turn this into THREE.OutlineShapeGeometry...?

Maybe. FWIW, the technique used in this PR is also used in http://threejs.org/examples/webgl_geometry_shapes.html.

koober committed:
Split shapes func. to shapes / lines; added holes

koober commented:
I looked at webgl_geometry_shapes before, but couldn't fit .generateShapes() in that format.
Anyway, I've updated that example to include holes, but also removed a couple of redundant lines.

WestLangley commented:
NIce.

https://rawgit.com/koober/three.js/dev/examples/webgl_geometry_shapes.html (proposed)

https://rawgit.com/mrdoob/three.js/dev/examples/webgl_geometry_shapes.html (original)

Maybe `webgl_geometry_shapes.html` should be in a separate PR, though.

koober committed:
Update webgl_geometry_shapes.html

koober committed:
Merge pull request #2 from koober/koober-patch-1

Update webgl_geometry_shapes.html

mrdoob commented:
Thanks!
